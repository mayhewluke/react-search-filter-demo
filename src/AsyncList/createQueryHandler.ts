import { merge, Observable, of, Subject } from "rxjs";
import { ajax } from "rxjs/ajax";
import {
  catchError,
  debounceTime,
  map,
  share,
  startWith,
  switchMap,
  withLatestFrom,
} from "rxjs/operators";

import {
  AsyncValue,
  AsyncValueTag,
  error,
  init,
  loading,
  success,
} from "common/async";

// Convenience function for pulling results out of an AsyncValue, or returning
// null if it was of a type that has no value (or if the results were null).
const valueOrNull = <T>(asyncValue: AsyncValue<T>) => {
  switch (asyncValue.state) {
    case AsyncValueTag.LOADING:
    case AsyncValueTag.SUCCESS:
      return asyncValue.results;
    default:
      return null;
  }
};

type Transform<T> = (data: any) => T;

const fetchResults = <T>(
  url: string,
  transform: Transform<T>,
): Observable<AsyncValue<any>> =>
  ajax.getJSON(url).pipe(
    map((response: any) => success(transform(response))),
    catchError(() => of(error<any>())),
  );

const createQueryHandler = <T>(transform: Transform<T>) => {
  const subject = new Subject<string>();

  // results$ needs to be shared or else it will be a cold observable - which
  // means that when loading$ subscribes to it (in addition to it being
  // subscribed to when merged into the output stream) it will run through its
  // pipeline again, causing another ajax request
  const results$ = subject.pipe(
    debounceTime(500), // TODO make time length configurable
    switchMap(url => fetchResults(url, transform)),
    share(),
  );

  const loading$ = subject.pipe(
    withLatestFrom(results$.pipe(startWith(init()))),
    map(([_, results]) => loading(valueOrNull(results))),
  );

  return {
    handler: subject.next.bind(subject),
    results$: merge(loading$, results$),
  };
};

export default createQueryHandler;
