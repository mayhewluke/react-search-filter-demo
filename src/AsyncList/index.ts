export {
  default as AsyncList,
  Meal,
  Props as AsyncListProps,
} from "./AsyncList";
export { default as createQueryHandler } from "./createQueryHandler";
