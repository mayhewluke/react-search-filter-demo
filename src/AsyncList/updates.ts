import { Meal } from "./reducer";

export enum UpdateTag {
  SEARCH_START = "SEARCH_START",
  FETCH_SUCCESS = "FETCH_SUCCESS",
  FETCH_FAIL = "FETCH_FAIL",
  RESET = "RESET",
}

export type Update =
  | { type: UpdateTag.SEARCH_START }
  | { type: UpdateTag.FETCH_SUCCESS; payload: Meal[] }
  | { type: UpdateTag.FETCH_FAIL }
  | { type: UpdateTag.RESET };

export const searchStart = (): Update => ({
  type: UpdateTag.SEARCH_START,
});

export const fetchSuccess = (meals: Meal[]): Update => ({
  payload: meals,
  type: UpdateTag.FETCH_SUCCESS,
});

export const fetchFail = (): Update => ({ type: UpdateTag.FETCH_FAIL });

export const reset = (): Update => ({ type: UpdateTag.RESET });
