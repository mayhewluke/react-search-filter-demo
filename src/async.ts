export enum AsyncValueTag {
  INIT = "INIT",
  LOADING = "LOADING",
  SUCCESS = "SUCCESS",
  ERROR = "ERROR",
}

export type AsyncValue<T> =
  | { state: AsyncValueTag.INIT }
  | { state: AsyncValueTag.LOADING; results: T | null }
  | { state: AsyncValueTag.SUCCESS; results: T }
  // Normally we would also want to have some kind of payload here, such as
  // `error: string`, but since we don't use that in this demo we omit it here.
  | { state: AsyncValueTag.ERROR };

// Convenience functions for constructing the above types, since TypeScript
// doesn't make working with tagged unions as easy as say, Haskell or Elm
export const init = <T>(): AsyncValue<T> => ({ state: AsyncValueTag.INIT });
export const loading = <T>(previousResults: T | null): AsyncValue<T> => ({
  results: previousResults,
  state: AsyncValueTag.LOADING,
});
export const success = <T>(results: T): AsyncValue<T> => ({
  results,
  state: AsyncValueTag.SUCCESS,
});
export const error = <T>(): AsyncValue<T> => ({
  state: AsyncValueTag.ERROR,
});
