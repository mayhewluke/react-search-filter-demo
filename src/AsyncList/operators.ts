import { Observable, of } from "rxjs";
import { ajax } from "rxjs/ajax";
import { catchError, distinctUntilChanged, map, pluck } from "rxjs/operators";

import { AsyncValue } from "async";

import { fetchFail, fetchSuccess, Update } from "./updates";

export interface Props {
  query: string;
  render: React.SFC<Meal>;
}

export type State = AsyncValue<Meal[]>;

export interface Meal {
  idMeal: string; // id
  strMeal: string; // Name
}

const url = (query: string) =>
  `https://www.themealdb.com/api/json/v1/1/search.php?s=${query}`;

export const fetchResults = (query: string): Observable<Update> =>
  ajax.getJSON(url(query)).pipe(
    map((response: any) => fetchSuccess(response.meals || [])),
    catchError(() => of(fetchFail())),
  );

// Can be used as a `pipe` operator
export const whenQueryChanges = (
  props$: Observable<Props>,
): Observable<string> =>
  props$.pipe(
    pluck<Props, Props["query"]>("query"), // extract the `query` property
    distinctUntilChanged(), // filter out consecutive duplicates
  );
