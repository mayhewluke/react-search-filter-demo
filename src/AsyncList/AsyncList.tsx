import React from "react";
import { FaSpinner } from "react-icons/fa";
import { Observable } from "rxjs";
import { map } from "rxjs/operators";

import { AsyncValue, caseOf } from "common/async";
import { componentFromStream } from "common/streamHelpers";

export interface Props {
  results: AsyncValue<Meal[]>;
  render: React.SFC<Meal>;
  errorComponent: JSX.Element;
  noResultsComponent: JSX.Element;
}

export interface Meal {
  idMeal: string; // id
  strMeal: string; // Name
}

const renderItems = (items: Meal[] | null, render: React.SFC<Meal>) => {
  if (items === null) {
    return null;
  }
  return items.map(render);
};

const AsyncList = componentFromStream((props$: Observable<Props>) => {
  return props$.pipe(
    map(({ results, render, errorComponent, noResultsComponent }) =>
      caseOf(results, {
        // TODO pass the error in so errorComponent can provide useful contextual info
        error: () => errorComponent,
        init: () => null,
        loading: items => (
          <div>
            <FaSpinner data-testid="loading" />
            {renderItems(items, render)}
          </div>
        ),
        success: items =>
          items.length > 0 ? renderItems(items, render) : noResultsComponent,
      }),
    ),
  );
});

export default AsyncList;
