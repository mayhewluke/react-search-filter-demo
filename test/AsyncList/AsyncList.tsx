import React from "react";
import { render } from "react-testing-library";

import { error, init, loading, success } from "common/async";

import { AsyncList, AsyncListProps as Props, Meal } from "AsyncList";

describe("AsyncList", () => {
  const renderFn = jest.fn();
  const errText = "This is an error message";
  const errorComponent = <span>{errText}</span>;
  const noResultsText = "No results found";
  const noResultsComponent = <span>{noResultsText}</span>;
  const defaultProps: Props = {
    errorComponent,
    noResultsComponent,
    render: renderFn,
    results: init(),
  };
  const component = (props: Partial<Props> = {}) => (
    <AsyncList {...defaultProps} {...props} />
  );
  const meals = [
    { idMeal: "123", strMeal: "Spicy Chicken" },
    {
      idMeal: "456",
      strMeal: "Rotisserie Chicken",
    },
  ];

  describe("when results are in the initial state", () => {
    it("renders nothing", () => {
      const { container } = render(component());

      expect(renderFn).not.toHaveBeenCalled();
      // Use a snapshot rather than asserting on number of child nodes - child
      // nodes may be empty and count as having rendered nothing but still throw
      // the test off
      expect(container).toMatchSnapshot();
    });
  });

  describe("when results are loading", () => {
    it("displays a loading indicator", () => {
      const results = loading<Meal[]>(null);
      const indicator = render(component({ results })).queryByTestId("loading");

      expect(indicator).not.toBeNull();
    });

    describe("with previous results available", () => {
      it("shows the previous results", () => {
        const results = loading(meals);
        render(component({ results }));

        expect(renderFn).toBeCalledTimes(2);
        expect(renderFn.mock.calls[0][0]).toEqual(meals[0]);
        expect(renderFn.mock.calls[1][0]).toEqual(meals[1]);
      });

      it("places the loading indicator above the previous results", () => {
        renderFn.mockImplementation(() => (
          <h1 key={Math.random()}>Previous results</h1>
        ));
        const results = loading(meals);
        const { container } = render(component({ results }));

        expect(container).toMatchSnapshot();
      });
    });
  });

  describe("when successful", () => {
    it("doesn't show the loading indicator", () => {
      const results = success([]);
      const indicator = render(component({ results })).queryByTestId("loading");

      expect(indicator).toBeNull();
    });

    it("doesn't show the errorComponent", () => {
      const results = success([]);
      const { queryByText } = render(component({ results }));

      expect(queryByText(errText)).toBeNull();
    });

    describe("with no results found", () => {
      it("renders the noResultsComponent", () => {
        const results = success([]);
        const { queryByText } = render(component({ results }));

        expect(queryByText(noResultsText)).not.toBeNull();
      });
    });

    describe("with results found", () => {
      it("doesn't render the noResultsComponent", () => {
        const results = success(meals);
        const { queryByText } = render(component({ results }));

        expect(queryByText(noResultsText)).toBeNull();
      });

      it("renders the results", () => {
        const results = success(meals);
        render(component({ results }));

        expect(renderFn).toBeCalledTimes(2);
        expect(renderFn.mock.calls[0][0]).toEqual(meals[0]);
        expect(renderFn.mock.calls[1][0]).toEqual(meals[1]);
      });
    });
  });

  describe("with an error", () => {
    it("renders the errorComponent", () => {
      const results = error<Meal[]>();
      const { queryByText } = render(component({ results }));

      expect(queryByText(errText)).not.toBeNull();
    });

    it("doesn't render the noResultsComponent", () => {
      const results = error<Meal[]>();
      const { queryByText } = render(component({ results }));

      expect(queryByText(noResultsText)).toBeNull();
    });

    it("doesn't show the loading indicator", () => {
      const results = error<Meal[]>();
      const { queryByTestId } = render(component({ results }));

      expect(queryByTestId("loading")).toBeNull();
    });
  });
});
