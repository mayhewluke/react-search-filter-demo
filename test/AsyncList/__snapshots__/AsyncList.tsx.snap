// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncList when results are in the initial state renders nothing 1`] = `<div />`;

exports[`AsyncList when results are loading with previous results available places the loading indicator above the previous results 1`] = `
<div>
  <div>
    <svg
      data-testid="loading"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
      />
    </svg>
    <h1>
      Previous results
    </h1>
    <h1>
      Previous results
    </h1>
  </div>
</div>
`;
